### DISCUSSION TITLE
How do I iterate this variable structure?

### DISCUSSION THREAD

**Discussion by Murphest** [nlohmann/json#2295] [Q&A] [answered]
I pared down this example JSON as much as I could while still retaining it's structure.  Can someone give me some pointers on how to iterate this?  I'm not sure how to handle having unknown array sizes at multiple levels (the geometry coords can be hundreds).

```json
{
    "type": "FeatureCollection",
    "features": [
    {
        "type": "Feature",
        "properties":
        {
            "NAME": "Arkansas",
            "CENSUSAREA": 52035.477000
        },
        "geometry":
        {
            "type": "Polygon",
            "coordinates": [
                [
                    [-94.042964, 33.019219],
                    [-94.042964, 33.019219]
                ]
            ]
        }
    },
    {
        "type": "Feature",
        "properties":
        {
            "NAME": "California",
            "CENSUSAREA": 155779.220000
        },
        "geometry":
        {
            "type": "MultiPolygon",
            "coordinates": [
                [
                    [
                        [-120.248484, 33.999329],
                        [-120.248484, 33.999329]
                    ]
                ],
                [
                    [
                        [-119.789798, 34.057260],
                        [-119.789798, 34.057260]
                    ]
                ]
            ]
        }
    }
    ]
}
```

**Comment by nlohmann** [comment, marked as answer]
Objects and arrays can both be access with `operator[]`. For an object, you pass the key; for an array, you pass the index.

Assuming you parsed the above JSON to a variable `j`, you can access the string "Arkansas" with `j["features"][0]["properties"]["NAME"]`.

To iterate, just use `begin()`/`end()` just like a `std::vector` or `std::map`. For easier access to keys and values in objects, there is the [`items`](https://nlohmann.github.io/json/doxygen/classnlohmann_1_1basic__json_a5961446010dfc494e0c247b4e9026977.html#a5961446010dfc494e0c247b4e9026977) method.

**Reply by Murphest** [reply]
I'm getting closer, but I think my old school thinking is getting in the way.   All I'm trying to pull out are the state name, and coordinates.  I haven't cracked the nut on the coordinates yet.  Also, I can almost guarantee I'm doing this "the hard way". 

	json features = j.at("features");
	int sz = features.size();
	int i,x;
	std::string name, geo_type;

	for (i = 0; i < sz; i++)
	{
		name = features[i]["properties"]["NAME"];
		geo_type = features[i]["geometry"]["type"];
		auto coords = features[i]["geometry"]["coordinates"][0];
		if (geo_type.compare("Polygon") == 0)
		{
			TRACE("%s : Polygon to follow\n", name.c_str());
			int nPairs = coords.size();
			for (x = 0; x < nPairs; x++)
			{
				auto thesecoords = coords[x];
			}
		}
		else if (geo_type.compare("MultiPolygon") == 0)
		{
			TRACE("%s : MultiPolygon to follow\n", name.c_str());
		}
		else
		{
			TRACE("%s : UNKNOWN GEOMETRY FOUND\n", name.c_str());
		}
	}

**Reply by Murphest** [reply]
Okay, it looks like I'm iterating properly now but it sure feels klunky and not elegant.

json features = j.at("features");
	int sz = features.size();
	int i,x;
	std::string name, geo_type;

	for (i = 0; i < sz; i++)
	{
		name = features[i]["properties"]["NAME"];
		geo_type = features[i]["geometry"]["type"];
		auto coords = features[i]["geometry"]["coordinates"][0];
		if (geo_type.compare("Polygon") == 0)
		{
			TRACE("%s : Polygon to follow\n", name.c_str());
			int nPairs = coords.size();
			for (x = 0; x < nPairs; x++)
			{
				auto thesecoords = coords[x];
				double lat = thesecoords[0].get<double>();
				double lon = thesecoords[1].get<double>();
				TRACE("   LAT/LON: %3.6f, %3.6f\n", lat, lon);

			}
		}
		else if (geo_type.compare("MultiPolygon") == 0)
		{
			int iSet, nSets, nPairs;
			TRACE("%s : MultiPolygon to follow\n", name.c_str());
			coords = features[i]["geometry"]["coordinates"][0][0];
			nSets = coords.size();

			for (iSet = 0; iSet < nSets; iSet++)
			{
				auto thisCoordSet = features[i]["geometry"]["coordinates"][iSet][0];
				nPairs = thisCoordSet.size();
				
				TRACE("    -------------------- NEW POLYGON -----------------\n");

				for (x = 0; x < nPairs; x++)
				{
					auto thesecoords = thisCoordSet[x];
					double lat = thesecoords[0].get<double>();
					double lon = thesecoords[1].get<double>();
					TRACE("   LAT/LON: %3.6f, %3.6f\n", lat, lon);

				}
			}
		}
		else
		{
			TRACE("%s : UNKNOWN GEOMETRY FOUND\n", name.c_str());
		}
	}

**Reply by nlohmann** [reply]
What about this?

```cpp
for (const auto& feature : j.at("features"))
{
    const std::string name = feature["properties"]["NAME"];
    const std::string geo_type = feature["geometry"]["type"];
    const auto& coords = feature["geometry"]["coordinates"];
    
    if (geo_type == "Polygon")
    {
        TRACE("%s : Polygon to follow\n", name.c_str());
        for (const auto& c : coords[0])
        {
            const auto lat = c[0].get<double>();
            const auto lon = c[1].get<double>();
            TRACE("   LAT/LON: %3.6f, %3.6f\n", lat, lon);
        }
    }
    else if (geo_type == "MultiPolygon")
    {
        TRACE("%s : MultiPolygon to follow\n", name.c_str());
        for (const auto& polygon : coords)
        {
            TRACE("    -------------------- NEW POLYGON -----------------\n");

            for (const auto& c : polygon[0])
            {
                const auto lat = c[0].get<double>();
                const auto lon = c[1].get<double>();
                TRACE("   LAT/LON: %3.6f, %3.6f\n", lat, lon);
            }
        }
    }
    else
    {
        TRACE("%s : UNKNOWN GEOMETRY FOUND\n", name.c_str());
    }
}
```

**Reply by Murphest** [reply]
Yes, substantially more elegant.  I don't understand how .at( ) returns an iterator for that for loop, but I'll look at the 'new' auto iterators, I wrote C++ for a couple decades (but not since  C++ 2.0) so those are new additions to the language for me.   Thanks Niels.

**Reply by nlohmann** [reply]
`at` returns a reference. The range `for` calls `begin()` and `end()` on it while iterating. It really makes that part of C++ quite nice!
